### ISSUE TITLE
Task goes back to SCHEDULED when it updates its own data

### ISSUE THREAD

**Issue by xesxen** [Netflix/conductor#1568]
In pull request #1377 there's a behavior change in the POST /tasks endpoint. When a job updates its own task (in my case: to add a log entry while it's still running), the task will switch back to SCHEDULED. Previously, the task kept its IN_PROGRESS status.

This was changed in WorkflowExecutor.java on the (added) line 811: https://github.com/Netflix/conductor/pull/1377/files#diff-a28de3002a75934bb7c693d03b674144R811

Is this an intentional change? If so, is there a way to bypass/disable this behavior?

**Comment by christianscheuer** [comment]
Doesn't it stay in IN_PROGRESS if you send this as part of the payload?
{
    "status": "IN_PROGRESS"
}

**Comment by xesxen** [comment]
No. If I fetch the ongoing task details and post the same unchanged JSON again to 'update' it (that JSON has the IN_PROGRESS state of course), it will switch to SCHEDULED

**Comment by christianscheuer** [comment]
Ah I see, thanks for the clarification. That indeed sounds like a really weird change. That might also explain why we're seeing multiple task workers receiving the same task to be processed (since the /ack endpoint has been deprecated but the /poll endpoint appears to allow multiple workers to process the same task).

**Comment by apanicker-nflx** [comment]
@xesxen @christianscheuer  This change was intentional, however the behavior of the task in the queue remains the same. 
[In the previous scenario](https://github.com/Netflix/conductor/pull/1377/files#diff-a28de3002a75934bb7c693d03b674144L851-L853), when the task was updated with `IN_PROGRESS` status, the task was removed from the queue and pushed back with an invisibility timeout of callbackAfterSeconds. This means that the task is now scheduled in the queue to be polled by a worker after the callbackAfterSeconds duration expires. When the status in such cases  
remained as IN_PROGRESS, it wasnt clear from the UI whether the task was being processed by a worker or was in invisible state in the queue. The change referenced in #1377 makes this explicit such that if the task status says IN_PROGRESS, it is actively processed by a worker whereas if the status is SCHEDULED, the task is in invisible state in the queue and will be available to be polled.

> /ack endpoint has been deprecated but the /poll endpoint appears to allow multiple workers to process the same task

The `ack` endpoint was deprecated in combination with removing the client-side queuing functionality. Additionally, the `poll` endpoint will `ack` the task [inherently](https://github.com/Netflix/conductor/blob/master/core/src/main/java/com/netflix/conductor/service/ExecutionService.java#L182) thus ensuring that only a single worker who successfully polled the task is actively processing it.

**Comment by christianscheuer** [comment]
Hi @apanicker-nflx.
Thank you for this explanation. I understand that the poll endpoint should inherently ack the task, but this appears to not be the case in real world usage with v2.25.5 when I last checked (perhaps this was fixed in recent changes?)

If I use [the C# client](https://github.com/courosh12/conductor-dotnet-client) with multiple threads in a single worker node using sub second polling intervals (for example 100ms), more than one thread will receive the task when querying the /poll endpoint, and therefore two or more threads start processing the task in parallel causing all sorts of issues.
This was on tag v2.25.5, using dynomite as persistence layer.

Wrt the /ack endpoint being deprecated - in current master, the Python client source code still uses it, and will not process any tasks right now since /ack on the server returns false:
https://github.com/Netflix/conductor/blob/master/client/python/conductor/ConductorWorker.py#L90

In other words - the Python client example ends up not processing any tasks, and the C# client will process the same task more than once.

Is there a canonical client implementation that is updated and guaranteed to work with the new server logic? Or could you point us in another direction as to why we would see the issue with the /poll endpoint returning the same task more than once for different threads?

**Comment by apanicker-nflx** [comment]
@christianscheuer I am not familiar with the C# client, but [one of the changes](https://github.com/Netflix/conductor/commit/b0b997f92b9630e527a9e17b864b03ecacbc48a7#diff-ccb60fa697fcb4a67b111665ca2abc91R101) that was made to the java client with the deprecation of the `ack` endpoint was that polls are now only made for a single task at a given time. The code that I [linked in the previous comment](https://github.com/Netflix/conductor/blob/master/core/src/main/java/com/netflix/conductor/service/ExecutionService.java#L182) addresses this point that a task is acked immediately upon polling and it is expected that the worker thread will start processing this task right away instead of queuing it again on the client. 

Thanks for pointing out the inconsistency with the python client, we will look into fixing this soon.

**Comment by christianscheuer** [comment]
@apanicker-nflx thank you for the info.

The code in [getLastPollTask](https://github.com/Netflix/conductor/blob/a2c37c5a14353603e460691c8aff7dd552af2477/core/src/main/java/com/netflix/conductor/service/ExecutionService.java#L182) - is that thread safe?
Meaning - if I have 10 Kubernetes nodes each with a worker calling the poll endpoint for the same task type - is there a lock on the call to `getLastPollTask` so that more than one request cannot retrieve the same task from this call:
```
List<Task> tasks = poll(taskType, workerId, domain, POLL_COUNT_ONE, POLLING_TIMEOUT_IN_MS);
```
Before they're acked a few lines further down?

It looks like a possible race condition if the calls to getLastPollTask are not synchronized - and this may explain why we're seeing more than one worker receiving the same task.

But perhaps it could have been related to the issue in this thread instead. We implemented "live" logging while a long-running task is running, so that the worker would update the task by setting the `logs` property of the task. If the change was made as described above so that IN_PROGRESS now becomes SCHEDULED when we try to update the log, then this may explain why the task gets rescheduled when all we wanted to say was that: "we're still processing but here is the latest additions to the log".
How would be a way to approach this logging while not rescheduling the task? Could there be made a change to the Update Task endpoint to allow us to update a task that we're still processing without re-scheduling it (which then would cause the issue with two workers working on it)?

**Comment by christianscheuer** [comment]
Pseudo code of what our worker was doing upon receiving a task:

```
//Start processing task
DoSomeInitialProcessing();
UpdateTask( /* send some logs here */ );
DoSomeMoreProcessing();
UpdateTask( /* send some more logs here */ );
DoTheLastBulkOfProcessing();
//finally, return to the server as being done with the task.
```

It appears if each call to update the task's logs will reschedule the task then we can't use this pattern.

**Comment by apanicker-nflx** [comment]
> But perhaps it could have been related to the issue in this thread instead. We implemented "live" logging while a long-running task is running, so that the worker would update the task by setting the logs property of the task. If the change was made as described above so that IN_PROGRESS now becomes SCHEDULED when we try to update the log, then this may explain why the task gets rescheduled when all we wanted to say was that: "we're still processing but here is the latest additions to the log".
How would be a way to approach this logging while not rescheduling the task? Could there be made a change to the Update Task endpoint to allow us to update a task that we're still processing without re-scheduling it (which then would cause the issue with two workers working on it)?

Based on this and the pseudo-code above, this is definitely the case. Instead of updating the task in this case, I would recommend using the `POST /tasks/{taskId}/log` API instead to append execution logs to the task, thereby keeping it IN_PROGRESS within your worker while also providing execution details for the task.

**Comment by christianscheuer** [comment]
Thank you so much for this. I suppose this also means that responseTime should be deprecated for task definitions?
https://netflix.github.io/conductor/tasklifecycle/

Or does a post to the log endpoint count as a response?

If we aren't supposed to update the task before it's done, it doesn't seem to make sense to me that there's both a responseTime and a timeout. Or is this to support tasks who reschedule themselves? In any case, the reason why we implemented our initial log post in the first time was actually to have a short responseTime (to let the server know yes, we're actually processing the task) and a long timeout (our tasks are fairly long running).

I guess this should be updated too
https://netflix.github.io/conductor/apispec/#schema-for-updating-task-result

Let me know if I can be of any help in updating this documentation and/or the python client to remove the ack'ing - however since I'm only starting to grasp the API changes I may need some help.

**Comment by apanicker-nflx** [comment]
@christianscheuer 
>Or does a post to the log endpoint count as a response?

This does not count as a response for now, but this counted as such. I understand the use case you mentioned for using the logs to keep short response timeouts and this would definitely help in that case. This change would need to be implemented however.

>is this to support tasks who reschedule themselves?

This is correct. [This](https://netflix.github.io/conductor/tasklifecycle/#timeout-seconds) explains how timeout and responseTimeout are used in tandem.

The schema for updating task result is still correct here since the worker responds with IN_PROGRESS status stating that the task is still in progress within the worker. The server upon receiving this response will mark this as SCHEDULED and enqueue the task with an invisibility of `callbackAfterSeconds`. After this time has elapsed, the worker will be able to poll for the task again and will receive the task with IN_PROGRESS status. So, from the perspective of the worker, the task has remained IN_PROGRESS.

However, having said that, the task lifecycle diagram needs to be updated and the documentation needs to be updated as well.

It would be great if you could make the corresponding changes to the python client. This would be a great help to the community.
